### ISSUE TITLE
Cannot update layers or sources when using MapboxMaps.MapView in UIViewRepresentable

### ISSUE THREAD

**Issue by jcfrk101** [mapbox/mapbox-maps-ios#1472]
## Environment

- Xcode version:
- 13.3
- iOS version:
- 15.4.1
- Devices affected:
- iPhone 12
- Maps SDK Version:
10.4.2
## Observed behavior and steps to reproduce
Add source and layer as show below 
```
let routeLine = PolylineAnnotation(lineCoordinates: routeCoordinates)
                    var source = GeoJSONSource()
                    source.data = .geometry(routeLine.geometry)
                    let sourceId = UUID().uuidString;
                    do{
                    try mapView.mapboxMap.style.addSource(source, id: sourceId)         
                        context.coordinator.sourceIds.append(source)
                        let layerId = "layer-"+UUID().uuidString
                        

                        var lineLayer = LineLayer(id:layerId)
                            lineLayer.source = sourceId
                            lineLayer.lineJoin = .constant(.bevel)
                            lineLayer.lineCap = .constant(.round)
                            lineLayer.lineWidth = .expression(
                            Exp(.interpolate) {
                                Exp(.linear)
                                Exp(.zoom)
                                1
                                1
                                10
                                2
                                13
                                4
                            }
                            )
                            lineLayer.lineColor = .constant(StyleColor(UIColor.alfaDarkGreen()))
                            context.coordinator.layerIds.append(lineLayer)
                          try mapView.mapboxMap.style.addPersistentLayer(lineLayer)
                    }
                    catch {
                        DDLogError("mapbox error -> "+error.localizedDescription)
                    }
```

attempt to call `try mapView.mapboxMap.style.updateLayer` later after updateUI is called. 
Get the following error `Layer layer-FD17FCA6-6113-4A30-B2B5-87C6C6CA6F3A` is not in style

## Expected behavior
layer should update.

**Comment by ZiZasaurus** [comment]
@jcfrk101 have you seen our SwiftUI example [here](https://github.com/mapbox/mapbox-maps-ios/blob/main/Apps/Examples/Examples/All%20Examples/Annotations/SwiftUIExample.swift)?

**Comment by jcfrk101** [comment]
Yes I have. And I am able to update and remove point annotations just fine. This only occurs with poly line annotations with layers and source.

**Comment by ZiZasaurus** [comment]
Thank you for the quick response. Looking into it!

**Comment by ZiZasaurus** [comment]
@jcfrk101 I am unable to reproduce this behavior. Can you please provide a sample application that reproduces this so we can see your implementation?

**Comment by ZiZasaurus** [comment]
Closing. Please feel free to reopen if you continue to experience this behavior.

**Comment by RheaParks** [comment]
I experience this issue as well, here's my representable:

```swift
struct WorkoutMap: UIViewRepresentable {
    
    // MARK: - Types
    
    private enum Track {
        case workout([CLLocationCoordinate2D])
        case hike([CLLocationCoordinate2D])
        
        var id: String {
            switch self {
            case .workout:
                return "workout"
            case .hike:
                return "hike"
            }
        }
        
        var coordinates: [CLLocationCoordinate2D] {
            switch self {
            case let .workout(coordinates):
                return coordinates
            case let .hike(coordinates):
                return coordinates
            }
        }
    }
    
    // MARK: - Parameters
    
    let workoutTrack: [CLLocationCoordinate2D]
    let hikeTrack: [CLLocationCoordinate2D]?
    
    // MARK: - Private Properties
    
    private let mapView: BaseMapView = {
        let mapView = BaseMapView()
        mapView.location.options.puckBearingEnabled = true
        mapView.location.options.puckType = .puck2D(.init(
            bearingImage: UIImage(named: "heading"),
            showsAccuracyRing: true))
        return mapView
    }()
    
    // MARK: - Representable Conformance
    
    typealias UIViewType = MapView
    
    func makeUIView(context: Context) -> MapView {
        mapView
    }
    
    func updateUIView(_ uiView: MapView, context: Context) {
        if let hikeTrack {
            updateTrack(.hike(hikeTrack))
        }
        updateTrack(.workout(workoutTrack))
    }
    
    // MARK: - Display tracks
    
    private func updateTrack(_ track: Track) {
        guard track.coordinates.count >= 2 else {
            // A track without 2 coordinates or more isn't actually a track.
            return
        }
            
        if (try? mapView.mapboxMap.style.source(withId: track.id)) != nil { // ❌❌❌❌❌ This always return nil. The other method `sourceExists` always return false
            // If the source exists, we merely udpate its data.
            // ❌❌❌❌❌ This never gets executed
            do {
                let geoJSON = GeoJSONObject.geometry(.lineString(.init(track.coordinates)))
                try mapView.mapboxMap.style.updateGeoJSONSource(withId: track.id, geoJSON: geoJSON)
            } catch {
                print(error.localizedDescription)
            }
        } else {
            // If the source doesn't exist, we need to add it…
            var geoJSONSource = GeoJSONSource()
            geoJSONSource.data = .geometry(.lineString(.init(track.coordinates)))
            do {
                try mapView.mapboxMap.style.addSource(geoJSONSource, id: track.id)
            } catch {
                print(error.localizedDescription)
            }
            
            // … Along with its layer
            if !mapView.mapboxMap.style.layerExists(withId: track.id) {
                var lineLayer = LineLayer(id: track.id)
                lineLayer.source = track.id
                // Set some style properties
                switch track {
                case .hike:
                    lineLayer.lineColor = .constant(.init(Constants.TrackColors.default))
                case .workout:
                    lineLayer.lineColor = .constant(.init(Constants.TrackColors.workout))
                }
                lineLayer.lineCap = .constant(.round)
                lineLayer.lineWidth = .constant(2)
                lineLayer.minZoom = Double(Constants.Zoom.focusedHike)
                // Add the line layer to the map.
                do {
                    try mapView.mapboxMap.style.addLayer(lineLayer)
                } catch {
                    print(error.localizedDescription)
                }
            }
        }
    }
    
}
```

Could we reopen this?

**Comment by RheaParks** [comment]
Ok! I've finally got it! 🥲 Actually, the issue does not come from Mapbox, but from how it's initialized in a UIViewRepresentable. It's an issue with how we use SwiftUI.

If your `UIViewRepresentable`, if you have parameters that are passed at init, it will redraw the representable each time one of those parameters change. In my example, each time `workoutTrack ` or `hikeTrack` was updated, the whole representable was redrawn.

However, UIViewRepresentable is smart and doesn't call `makeUIView(context:)` twice! BUT, everything that was declared outside of `makeUIView(context:)` gets indeed replaced. In my case, the `mapView` of my representable was initialized _with_ the representable. Whereas I should have initialized it inside `makeUIView(context:)` and then pass the reference on each update using the `uiView` provided by `func updateUIView(_ uiView: MapView, context: Context)`.

I just modified my code this way, and now everything works fine! I can't believe I didn't know this, it might even explain other issues I've encountered! 😀
